tool/tsh/common/tsh.go

```
@@ -2555,49 +2555,73 @@ func onLogout(cf *CLIConf) error {
 
 	switch {
 	// Proxy and username for key to remove.
-	case proxyHost != "" && cf.Username != "":
-		tc, err := makeClient(cf)
+	case proxyHost != "":
+
+		// In the event --user flag is not supplied, and there is only one identity,
+		// we can simply log out the single identity.
+		// Extract the list of identities (usernames) currently in keystore.
+		clientStore := cf.getClientStore()
+		usernames, err := clientStore.GetIdentities(proxyHost)
 		if err != nil {
 			return trace.Wrap(err)
 		}
-		// Load profile for the requested proxy/user.
-		profile, err := tc.ProfileStatus()
-		if err != nil && !trace.IsNotFound(err) && !trace.IsCompareFailed(err) {
-			return trace.Wrap(err)
+		if cf.Username == "" {
+			logger.DebugContext(cf.Context, "No --user flag provided, but identities found for proxy",
+				"proxy_host", proxyHost,
+				"users", usernames)
+
+			if len(usernames) == 1 {
+				cf.Username = usernames[0]
+			}
 		}
 
-		// Log out user from the databases.
-		if profile != nil {
-			for _, db := range profile.Databases {
-				logger.DebugContext(cf.Context, "Logging user out of database",
-					"user", profile.Name,
-					"database", db,
-				)
-				err = dbprofile.Delete(tc, db)
-				if err != nil {
-					return trace.Wrap(err)
+		switch {
+		case cf.Username != "":
+			tc, err := makeClient(cf)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			// Load profile for the requested proxy/user.
+			profile, err := tc.ProfileStatus()
+			if err != nil && !trace.IsNotFound(err) && !trace.IsCompareFailed(err) {
+				return trace.Wrap(err)
+			}
+
+			// Log out user from the databases.
+			if profile != nil {
+				for _, db := range profile.Databases {
+					logger.DebugContext(cf.Context, "Logging user out of database",
+						"user", profile.Name,
+						"database", db,
+					)
+					err = dbprofile.Delete(tc, db)
+					if err != nil {
+						return trace.Wrap(err)
```

I know that you didn't change this, but I wonder if this is the correct behavior.

If we fail to delete one of the DB profiles, should we return an error and leave the user in a half-logged-out state? Or should we instead continue to log out as best we can?

---

It seems other areas of onLogout contain this behavior as well. There may be extra overhead in displaying to the user which profiles were successfully logged out or not, and what actions should be taken for those that weren't. Shall this be looked at in another issue?

---

Yep, sounds good. Feel free to create a new issue to track this.